### ISSUE TITLE
Pool mapping

### ISSUE THREAD

**Issue by regbo** [bbottema/generic-object-pool#3]
As mentioned from another issue I came here from Stormpot. I have a resource that takes a while to spin up, and it's best done in parallel. It looks like your project does it all on one thread.

To overcome this I created two wrapper classes below. The "map" methods allow me to take something like a CompletableFuture, and make sure all claims from a pool are mapped. That way I can allocate a future and block elsewhere if need be.

If possible, I think adding a method like this to your GenericObjectPool class would be great:

````java
default <U> Pool<U, X> map(Function<? super X, U> mapper);
````

Here are my wrappers (both use immutables):

````java
@Value.Immutable
@Value.Style(stagedBuilder = true)
public interface Pool<T, X> {

	GenericObjectPool<X> getGenericObjectPool();

	Function<? super X, T> getMapper();

	default <U> Pool<U, X> map(Function<? super X, U> mapper) {
		return ImmutablePool.<U, X>builder().genericObjectPool(getGenericObjectPool()).mapper(mapper).build();
	}

	/**
	 * @param timeout
	 * @return
	 * @throws InterruptedException
	 * @throws IllegalStateException
	 * @see org.bbottema.genericobjectpool.GenericObjectPool#claim(org.bbottema.genericobjectpool.util.Timeout)
	 */
	@SuppressWarnings("resource")
	@Nullable
	default Poolo<T> claim(Timeout timeout) throws InterruptedException, IllegalStateException {
		var po = getGenericObjectPool().claim(timeout);
		if (po == null)
			return null;
		var poolo = new Poolo<X>() {

			@Override
			public PoolableObject<?> getPoolableObject() {
				return po;
			}

			@Override
			public X getAllocatedObject() {
				return po.getAllocatedObject();
			}
		};
		return poolo.map(getMapper());
	}

	public static <X> ImmutablePool<X, X> build(GenericObjectPool<X> genericObjectPool) {
		return build(genericObjectPool, Function.identity());
	}

	public static <T, X> ImmutablePool<T, X> build(GenericObjectPool<X> genericObjectPool,
			Function<? super X, T> mapper) {
		return ImmutablePool.<T, X>builder().genericObjectPool(genericObjectPool).mapper(mapper).build();
	}

	public static abstract class Abs<T, X> implements Pool<T, X> {

		private final GenericObjectPool<X> genericObjectPool;

		public Abs(GenericObjectPool<X> genericObjectPool) {
			super();
			this.genericObjectPool = Objects.requireNonNull(genericObjectPool);
		}

		@Override
		public GenericObjectPool<X> getGenericObjectPool() {
			return genericObjectPool;
		}
	}

	public static class Impl<X> extends Pool.Abs<X, X> {

		public Impl(GenericObjectPool<X> genericObjectPool) {
			super(genericObjectPool);
		}

		@Override
		public Function<? super X, X> getMapper() {
			return Function.identity();
		}

	}

	// *****************************delegates

````

And

````java
@Value.Immutable
@Value.Style(stagedBuilder = true)
public interface Poolo<T> extends AutoCloseable {

	PoolableObject<?> getPoolableObject();

	/**
	 * @return
	 * @see org.bbottema.genericobjectpool.PoolableObject#getAllocatedObject()
	 */
	@Nullable
	T getAllocatedObject();

	@SuppressWarnings("unchecked")
	default <U> Poolo<U> map(Function<? super T, U> mapper) {
		Objects.requireNonNull(mapper);
		var allocatedObject = getAllocatedObject();
		var allocatedObjectUpdated = mapper.apply(allocatedObject);
		if (allocatedObject == allocatedObjectUpdated)
			return (Poolo<U>) this;
		return ImmutablePoolo.<U>builder()
				.poolableObject(getPoolableObject())
				.allocatedObject(allocatedObjectUpdated)
				.build();
	}

	@Override
	default void close() {
		this.getPoolableObject().release();
	}

	// **********************delegates
````

**Comment by regbo** [comment]
Wow... might have wasted a lot of time here. Just saw this project:

https://github.com/bbottema/clustered-object-pool
